### PR TITLE
GitHub-Issue#2778: Added CloudWatchLogs Buffer, ThresholdCheck, and ClientFactory utilities.

### PR DESCRIPTION
**Pull request by MaGonzalMayedo**
### Description
This change contains the Buffer, ThresholdCheck, and ClientFactory implementations for the CloudWatchLogs Sink. It provides to add ability for threshold checking, AWS Client Interfacing and customer support for threshold checks.
 
### Issues Resolved
This PR will work towards resolving CWL-Sink for Data-Prepper [Issue https://github.com/https://github.com/opensearch-project/data-prepper/issues/2778]
Link: https://github.com/opensearch-project/data-prepper/issues/2778#issue-1731228391
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ * limits of a buffer and CloudWatchLogsClient before making a
+ * PutLogEvent request to AWS.
+ */
+public class ThresholdCheck {
+    private final int batchSize;
+    private final int maxEventSize;
+    private final int maxRequestSize;
+    private final int logSendInterval;
+
+    ThresholdCheck (int batchSize, int maxEventSize, int maxRequestSize, int logSendInterval) {
```
Comment (asifsmohammed) [inline data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java:21]: All the parameters should be final here and the below methods

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactory.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        final AwsCredentialsProvider awsCredentialsProvider = awsCredentialsSupplier.getProvider(awsCredentialsOptions);
+
+        return CloudWatchLogsClient.builder()
+                .region(awsConfig.getAwsRegion())
+                .credentialsProvider(awsCredentialsProvider)
+                .overrideConfiguration(createOverrideConfiguration(awsConfig)).build();
+    }
+
+    private static ClientOverrideConfiguration createOverrideConfiguration(final AwsConfig awsConfig) {
+        final RetryPolicy retryPolicy = RetryPolicy.builder().numRetries(AwsConfig.DEFAULT_CONNECTION_ATTEMPTS).build();
```
[the remark this answers is not in the document]
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactory.java:40]: Ohhh my apologies, I could have gotten the two parameters confused, so we are to have a parameter called "DEFAULT_RETRY_COUT" which is to be used within the Sink itself for retryable exceptions, while the "DEFAULT_CONNECTION_ATTEMPTS" is meant to be for the SDK retry.

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/Buffer.java
[the remark this answers is not in the document]
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/Buffer.java:1]: The s3-sink buffer does have most of the methods we have in the CloudWatchLogs, but the main issue is the coupling that happens in the s3 version with the flushToS3. If it didn't have that method it could be moved to a commons for general purpose.

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java (hunk 1 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java]: I have made the parameters final for the next revision.

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBuffer.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.buffer;
+
+import java.util.ArrayList;
+
+public class InMemoryBuffer implements Buffer {
```
Comment (kkondaka) [inline data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBuffer.java:10]: I think you need flush API to flush the buffer contents to Cloud Watch, right?

Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBuffer.java:10]: I separated the dependency of the flushing from the Buffer, the buffer is purely for holding information. The next PR will include a CloudWatchLogsClientService class which uses the SDK and Buffer to flush to CloudWatchLogs. Is this what you meant or was it some other flush API?

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.threshold;
+/**
+ * ThresholdCheck receives paramaters for which to reference the
+ * limits of a buffer and CloudWatchLogsClient before making a
+ * PutLogEvent request to AWS.
+ */
+public class ThresholdCheck {
```
Comment (dlvenable) [inline data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java:12]: This class appears to be unused (aside from unit tests). Why does it have so many public methods? Will the calling code need this level of granularity?

For example, why do we need to know equal versus greater than? Why not combine them?

Also, why does the caller need to know which threshold is reached (e.g. bytes vs. interval vs. event size)?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/Buffer.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * Size of buffer in bytes.
+     * @return int
+     */
+    int getBufferSize();
+
+    void writeEvent(byte[] event);
+
+    byte[] getEvent();
```
[the remark this answers is not in the document]
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/Buffer.java:30]: Thank you, I have renamed the method. The main idea is that the buffer, must follow a push and pop approach that is generalized in bytes (Similar to a Queue), as we need to process the log events in order of arrival.

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java (hunk 4 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java:12]: The main reason for the granularity is for the limits for sending data through the CloudWatchLogs SDK, as the API has byte size limitations and to be as strict with the calls to it, we need to be able to define a hard threshold (with byte size and batch size). This way the customer has control over how large the request to the API can be.

File: data-prepper-plugins/cloudwatch-logs/build.gradle [hunk 6]
```diff
@@ -10,6 +10,8 @@ repositories {
 dependencies {
     api project(':data-prepper-api')
     implementation project(':data-prepper-plugins:aws-plugin-api')
+    testImplementation platform('org.junit:junit-bom:5.9.1')
```
[the remark this answers is not in the document]
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/build.gradle:13]: I have made some changes to the build.gradle file to include catalogs, but the one implementation that I believe I need is the aws-plugin-api, as I tested without it and it did not work.

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java (hunk 4 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java:12]: I have made most of the internal methods private as they are not actually necessary to the caller, thank you. The only ones that are public are checkMaxEventSize(), as this is more of an error checking method that is used prior to thresholding.

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactory.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactory.java:1]: Let's rename this file to not use abbreviations if possible.

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        return ((checkEqualBatchSize(batchSize) || checkEqualMaxRequestSize(currentRequestSize)) && (batchSize > 0));
+    }
+
+    /**
+     * Checks if the interval passed in is equal to or greater
+     * than the threshold interval for sending PutLogEvents.
+     * @param currentTimeSeconds int denoting seconds.
+     * @return boolean - true if greater than or equal to logInterval, false otherwise.
+     */
+    private boolean checkLogSendInterval(final long currentTimeSeconds) {
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java:53]: nit: would renaming the `check_` methods to match their description and return types. It sounds like a validation method but its just a comparison.

For example `checkGreaterThanMaxEventSize` -> `isGreaterThanMaxEventSize`

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.buffer;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.event.JacksonEvent;
+import org.opensearch.dataprepper.model.record.Record;
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java:12]: Record is deprecated.

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        assertThat(inMemoryBuffer.getBufferSize(), equalTo(42));
+    }
+
+    @Test
+    void check_buffer_has_right_size_test() {
+        for (Record<Event> eventToTest: getTestCollection()) {
+            inMemoryBuffer.writeEvent(eventToTest.getData().toJsonString().getBytes());
+        }
+
+        assertThat(inMemoryBuffer.getBufferSize(), equalTo(63));
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java:84]: Where does 63 come from?

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactory.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        final AwsCredentialsOptions awsCredentialsOptions = convertToCredentialOptions(awsConfig);
+        final AwsCredentialsProvider awsCredentialsProvider = awsCredentialsSupplier.getProvider(awsCredentialsOptions);
+
+        return CloudWatchLogsClient.builder()
+                .region(awsConfig.getAwsRegion())
+                .credentialsProvider(awsCredentialsProvider)
+                .overrideConfiguration(createOverrideConfiguration(awsConfig)).build();
+    }
+
+    private static ClientOverrideConfiguration createOverrideConfiguration(final AwsConfig awsConfig) {
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactory.java:38]: `awsConfig` is unused

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.threshold;
+/**
+ * ThresholdCheck receives paramaters for which to reference the
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java:8]: typo

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java (hunk 9 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java:12]: Record is deprecated, but still used as part of Data Prepper. I mainly used it as the other plugins use it as standard for output method calls.
@kkondaka @dinujoh What instance should be used in its stead or will this be added later on?

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/configuration/CwlSinkConfigTest.java [hunk 13]
```diff
@@ -1,3 +1,8 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
 package org.opensearch.dataprepper.plugins.sink.configuration;
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/configuration/CwlSinkConfigTest.java:6]: `test` and `main` paths should be the same
`org.opensearch.dataprepper.plugins.sink.configuration`
`org.opensearch.dataprepper.plugins.sink.config;`

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferFactory.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.buffer;
+
+public class InMemoryBufferFactory implements BufferFactory{
+    @Override
+    public Buffer getBuffer() {
+        return new InMemoryBuffer();
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferFactory.java:11]: No test coverage here. Is this going to be covered in another PR?

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheckTest.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        assertThat(thresholdCheck.isGreaterThanThresholdReached(ThresholdConfig.DEFAULT_LOG_SEND_INTERVAL_TIME -1, ThresholdConfig.DEFAULT_SIZE_OF_REQUEST, batch_size), is(false));
+    }
+
+    @Test
+    void check_equal_than_threshold_conditions_requeset_size_true() {
+        assertThat(thresholdCheck.isEqualToThresholdReached(ThresholdConfig.DEFAULT_SIZE_OF_REQUEST, ThresholdConfig.DEFAULT_BATCH_SIZE - 1), is(true));
+    }
+
+    @Test
+    void check_equal_than_threshold_conditions_requeset_size_false() {
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheckTest.java:81]: typo

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java (hunk 8 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java:53]: Got it, I have made the changes for the next PR update.

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactory.java (hunk 7 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactory.java:1]: I have renamed the files according to "CloudWatchLogs" prefix.

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheckTest.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+    @BeforeEach
+    void setUp() {
+        thresholdCheck = new ThresholdCheck(ThresholdConfig.DEFAULT_BATCH_SIZE, ThresholdConfig.DEFAULT_EVENT_SIZE,
+                ThresholdConfig.DEFAULT_SIZE_OF_REQUEST, ThresholdConfig.DEFAULT_LOG_SEND_INTERVAL_TIME);
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {55, 80, 100})
+    void check_max_event_size_invalid(final int event_size) {
+        assertThat(thresholdCheck.checkGreaterThanMaxEventSize(event_size), is(true));
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheckTest.java:30]: Can be changed to assertTrue.
Lets also try to move `checkGreaterThanMaxEventSize()` out of the assert statement for better readability.

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactoryTest.java [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+        final CloudWatchLogsClient cloudWatchLogsClientToTest = CwlClientFactory.createCwlClient(awsConfig, awsCredentialsSupplier);
+
+        assertThat(cloudWatchLogsClientToTest, notNullValue());
+    }
+
+    @Test
+    void check_created_client_with_no_params() {
+        final CloudWatchLogsClient cloudWatchLogsClient = CwlClientFactory.createCwlClient(awsConfig, awsCredentialsSupplier);
+
+        assertThat(cloudWatchLogsClient, notNullValue());
```
Comment (markkuhn) [inline data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactoryTest.java:57]: can be changed to `assertNotNull`

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java (hunk 10 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java:84]: Sorry, I have standardized it into a method for simplicity. 63 is supposedly the size of a single event json string by the number of those event copies (3).

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java (hunk 12 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheck.java:8]: Fixed the typo.

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/configuration/CwlSinkConfigTest.java (hunk 13 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/configuration/CwlSinkConfigTest.java:6]: I have made the change to config for both of these sub-directories.

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferFactory.java (hunk 14 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferFactory.java:11]: I have added the test coverage.

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheckTest.java (hunk 16 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/threshold/ThresholdCheckTest.java:30]: I have made changes to make the code more readable.

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactoryTest.java (hunk 17 above)
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/client/CwlClientFactoryTest.java:57]: Made the change

File: data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/Buffer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/Buffer.java]: But, why load them one at a time? Shouldn't you get a certain threshold amount each time?

[the remark this answers is not in the document]
Comment (MaGonzalMayedo) [reply data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/buffer/Buffer.java]: The calling code would check if the buffer has reached any thresholds, and then iterate through it, getting an event at a time and wrapping it around a InputLogEvent data structure. These InputLogEvents would be placed in a list inside a PutLogEventRequest which would then be sent to CloudWatchLogs as a request.

**Comment by dlvenable** [comment]
@MaGonzalMayedo , This looks good. There are currently checkstyle errors:

```
Error: eckstyle] [ERROR] /home/runner/work/data-prepper/data-prepper/data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferFactoryTest.java:11:36: Using the '.*' form of import should be avoided - org.hamcrest.Matchers.*. [AvoidStarImport]
Error: eckstyle] [ERROR] /home/runner/work/data-prepper/data-prepper/data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsClientFactoryTest.java:27:15: Unused import - org.hamcrest.Matchers.notNullValue. [UnusedImports]
```

**Review by dlvenable — APPROVED** [review]
Nice! Thanks for this contribution!

**Review by markkuhn — APPROVED** [review]
🚀

File: data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java (hunk 9 above)
Comment (kkondaka) [reply data-prepper-plugins/cloudwatch-logs/src/test/java/org/opensearch/dataprepper/plugins/sink/buffer/InMemoryBufferTest.java:12]: It's OK to use Record. It is my understanding It is deprecated in using the metadata but not entirely. @dlvenable, why is Record marked deprecared?

**Review by kkondaka — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]
